### PR TITLE
Proposal: Change --publish=SPEC to allow binding to custom host port ranges

### PR DESCRIPTION
**Pull request by dkjer**
This PR is similar to #12622, but is more narrow and focuses purely on the publish SPEC.  This change would allow specifying a range for the hostPort of a publish SPEC (iff containerPort is not a range), which would override the default binding to ephemeral ports and instead bind within the specified hostPort range:

Example:

```
$ docker run -d -p 8000-9000:5000 training/webapp python app.py
```

This would bind port 5000 in the container to a randomly available port between 8000 and 9000 on the host.

### CODE REVIEW

**Comment by bru7us** [comment]
It looks like there are a few conflicts between this and #12622, I think the better solution would be some combination of both PRs (see: https://github.com/docker/docker/pull/10052#issuecomment-93592231) - I just never got around to implementing this format.

A couple of questions for you:
- What does the hostConfig look like on a container that is started with this feature?
- What happens on restart of the container started this way? (try to keep the same allocated port, or always re-allocate from the range?)

I love what you've done with the port range keys in the global map. I started out doing something similar, but with multiple port maps so I came across an issue where you can have potentially overlapping maps that will conflict with each other. This is a much more elegant approach.

**Comment by dkjer** [comment]
@bru7us 

Here is an example session:
```
$ docker run -d -p 8000-9000:80 --name porttest nginx
$ docker port porttest
80/tcp -> 0.0.0.0:8003
$ docker inspect porttest
[{
  ...
    "HostConfig": {
  ...
        "PortBindings": {
            "80/tcp": [
                {
                    "HostIp": "",
                    "HostName": "",
                    "HostPort": "8000-9000"
                }
            ]
        },
  ...
    },
  ...
    "NetworkSettings": {
  ...
        "Ports": {
            "443/tcp": null,
            "80/tcp": [
                {
                    "HostIp": "0.0.0.0",
                    "HostName": "0.0.0.0",
                    "HostPort": "8003"
                }
            ]
        }
  ...
    },
  ...
}
]
$ docker restart porttest
$ docker port porttest
80/tcp -> 0.0.0.0:8004
```

The restart does not attempt to keep the same port, but it does re-allocate from within the same requested port range.

**Comment by bru7us** [comment]
Thanks for the follow up - I'm struggling to understand the benefit of using this instead of just any ephemeral port (outside of the scope of the other PRs). If every restart auto-allocates a port - why not just use the ephemeral range?

I think my latest comment (https://github.com/docker/docker/pull/12622#issuecomment-98555858) sums up the ultimate solution (a combination of your work, mine and that in #10052), do you have any feedback on that?

**Comment by dkjer** [comment]
My use-case is that I'm attempting to use AWS Security Groups to differentiate between 'public' vs 'private' port ranges to allow external traffic.  I need to be able to assign container ports in a more controlled manner than just using ephemeral ports.

I'll take a look at #12622 and provide some feedback

**Comment by LK4D4** [comment]
@docker/core-maintainers for design review. I think it is smaller and less-breaking than #12622

**Comment by bru7us** [comment]
> I think it is smaller and less-breaking than #12622

It's definitely smaller, but it's also different than #12622 as it doesn't provide persistence for dynamic allocations (which is the primary goal of that PR - I just absorbed the changes from this one into it for better use case coverage).

**Comment by dkjer** [comment]
Rebased off master.  Updated vendor/src/github.com/docker/libnetwork to match https://github.com/docker/libnetwork/pull/285

**Comment by icecrime** [comment]
@mavenugo @mrjana @aboch Thoughts?

**Comment by dkjer** [comment]
@mavenugo @mrjana @aboch @icecrime  Any word on this?

**Comment by icecrime** [comment]
*Group review session*

Design LGTM assuming that the container doesn't persist the elected port (which was the reason that the alternative implementation got previously closed).

Ping @mrjana @mavenugo @aboch: please review!

**Comment by dkjer** [comment]
@mrjana @mavenugo @aboch @icecrime  : Rebased off master

**Comment by mavenugo** [comment]
@dkjer Didn't get a chance to review the code in depth yet.
But in order to parallelize, a high level comment is to please open a PR against https://github.com/docker/libnetwork project for all the changes that you have made under `vendor/src/github.com/docker/libnetwork/` and lets vendor it in docker/docker.
Since the code is available here, we will review the content in parallel.

**Comment by mavenugo** [comment]
@dkjer have a comment / question for your PR in libnetwork https://github.com/docker/libnetwork/pull/285. Can you please address that so that we can get this PR moving ?

**Comment by dkjer** [comment]
@mavenugo Sure, apologies for the delay

**Comment by mavenugo** [comment]
@dkjer latest libnetwork with your patch is in docker master now. Can you please rebase to the latest and update this PR ?

**Comment by dkjer** [comment]
@mavenugo  Rebased

File: pkg/nat/nat.go [hunk 1]
```diff
@@ -59,6 +62,21 @@ func ParsePort(rawPort string) (int, error) {
...
+func ParsePortRange(rawPort string) (int, int, error) {
+	if len(rawPort) == 0 {
+		return 0, 0, nil
+	}
+	start, end, err := parsers.ParsePortRange(rawPort)
+	if err != nil {
+		return 0, 0, err
+	}
+	if end < start {
+		return 0, 0, fmt.Errorf("Invalid port range: %d-%d", start, end)
```
Comment (mavenugo) [inline pkg/nat/nat.go]: Minor nit : https://github.com/golang/go/wiki/CodeReviewComments#error-strings

**Comment by bru7us** [comment]
@mavenugo, did you see docker/libnetwork#453 yet? If we can catch it before this PR gets merged we can easily maintain full backward compatibility in libnetwork with a minor update to this PR to include separate start port and specified port.

If this gets merged first, there will be a dependency on docker and libnetwork to be updated together to get persistence policies in.

**Comment by dkjer** [comment]
@mavenugo Error string updated / commit squashed

File: pkg/nat/nat.go [hunk 2]
```diff
@@ -84,6 +102,14 @@ func (p Port) Int() int {
 	return int(port)
 }
 
+// Range returns the start/end port numbers of a Port range as ints
+func (p Port) Range() (int, int) {
+	// We don't need to check for an error because we're going to
+	// assume that any error would have been found, and reported, in NewPort()
```
Comment (mavenugo) [inline pkg/nat/nat.go]: Is there a reason why we are trying to ignore the error (other than reason mentioned here?).
I guess we could break this assumption with a simple Unit-Test.
Though it is not something we will see today, this assumption can lead to issues down the line.
Hence it would be better to not ignore the error.

File: pkg/nat/nat.go [hunk 3]
```diff
@@ -59,6 +62,21 @@ func ParsePort(rawPort string) (int, error) {
...
+// ParsePortRange parses the port range string and returns start/end ints
+func ParsePortRange(rawPort string) (int, int, error) {
+	if len(rawPort) == 0 {
+		return 0, 0, nil
+	}
+	start, end, err := parsers.ParsePortRange(rawPort)
+	if err != nil {
+		return 0, 0, err
+	}
+	if end < start {
```
Comment (aboch) [inline pkg/nat/nat.go]: Looks like `parser.ParsePortRange()` already checks for `end < start`.
Even if you want to override the error message with a more specific one, this code would not be executed.

File: pkg/nat/nat.go [hunk 4]
```diff
@@ -84,6 +102,14 @@ func (p Port) Int() int {
 	return int(port)
 }
 
+// Range returns the start/end port numbers of a Port range as ints
+func (p Port) Range() (int, int) {
+	// We don't need to check for an error because we're going to
+	// assume that any error would have been found, and reported, in NewPort()
+	portStart, portEnd, _ := ParsePortRange(p.Port())
```
Comment (aboch) [inline pkg/nat/nat.go]: I agree with @mavenugo, `Range()` is an exported function and `type Port` can be constructed in an unsafe way.

**Comment by aboch** [comment]
@dkjer Could you please cover this in current integration cli test for port specs: `dcoker_cli_port_test.go`
You can just integrate the existing test functions with few more cases where the host port range is specified (correctly and incorrectly) and verify it.

File: pkg/nat/nat.go (hunk 2 above)
Comment (dkjer) [inline pkg/nat/nat.go]: @mavenugo I agree.  This was just following the pattern given in Port.Int(), which ignores errors with the same assumption.  I will add the error case

**Comment by dkjer** [comment]
@mavenugo @aboch Addressed code review comments and added integration test.  Squashed commit

**Comment by aboch** [comment]
LGTM

**Comment by mavenugo** [comment]
@bru7us sorry not yet. we have start with design review of https://github.com/docker/libnetwork/pull/453 and make sure the features align with all the upcoming features. We have to also look @ all the design review comments of #12622. I think it is okay to keep 2 separate.

**Comment by bru7us** [comment]
Ok, thanks for the response.

**Comment by mavenugo** [comment]
@dkjer Thanks for addressing the comments.
The implementation LGTM and is inline with the use-case that was discussed in this PR during the design review.

But, I think we need some inputs from the docs team especially because the `-p <host-port-range>:<container-port-range>` behaves a bit different from the newly introduced `-p <host-port-range>:<container-port> and also the persistence expectations are a bit different.

ping @moxiegirl @thaJeztah

**Comment by moxiegirl** [comment]
Thanks @dkjer LGTM

File: docs/reference/run.md [hunk 5]
```diff
@@ -984,6 +984,7 @@ or override the Dockerfile's exposed defaults:
                    format: ip:hostPort:containerPort | ip::containerPort | hostPort:containerPort | containerPort
                    Both hostPort and containerPort can be specified as a range of ports.
                    When specifying ranges for both, the number of container ports in the range must match the number of host ports in the range. (e.g., `-p 1234-1236:1234-1236/tcp`)
+                   When specifying a range for hostPort only, the containerPort must not be a range.  In this case the container port is published somewhere within the specified hostPort range. (e.g., `-p 1234-1236:1234/tcp`)
```
Comment (moxiegirl) [inline docs/reference/run.md]: @dkjer Thank you for your contribution.  This is great information and I think expanding it will help readers get this information more easily.  Also, I'd like it if we could put this information in a table.  The table is beyond your scope, if you are willing to add it, good. If not I can carry the work for you.

--> https://gist.github.com/moxiegirl/8fa6820db70a011a8414

<table>
  <tr>
    <th>Option</th>
    <th>Description</th>
  </tr>
  <tr>
    <td><code>--expose=[]</code></td>
    <td>Expose a port or a range of ports from the container,without publishing it to your host</td>
  </tr>
  <tr>
    <td><code>-P=false</code></td>
    <td>Publish all exposed ports to the host interfaces.</td>
  </tr>
  <tr>
    <td><code>-p=[]<code></td>
    <td>
      <p>
      Publish a container᾿s port or a range of ports to the host. The value's format can be any of:    
      <ul>
      <li> <code>ip:hostPort:containerPort</code> 
      <li> <code>ip::containerPort</code> 
      <li> <code>hostPort:containerPort</code>
      <li> <code>containerPort</code>
      </ul>
      </p>
      <p>
      Use the <code>ip:hostPort:containerPort</code> or <code>ip::containerPort</code> formats
      if you want to be sure to bind a specific interface on the host side.
      </p>
      <p>
      You can specify <i>both</i> the <code>hostPort</code> and <code>containerPort</code> as port ranges.
      When specifying ranges for both, the number of values in the ranges
      must match, for example:<code>-p 2000-3000:2500-3500/tcp</code> The total number of
      values are both 1000.
      </p>
      <p>
      You can also specify a range for <code>hostPort</code> alone and a single
      value for the container:<code>-p 1234-1236:2000/tcp</code> In this case,
      the <code>containerPort</code> is mapped to a value somewhere within the
      given <code>hostPort</code> range. Use the <code>docker port</code>
      command to see the actual mapping.
      </p>
      </td>
  </tr>
  <tr>
    <td><code>--link=""</code></td>
    <td>Add link to another container (<code>id:alias</code>, <code>name:alias</code>, <code>name</code>, <code>id</code>)</td>
  </tr>
</table>

Comment (dkjer) [inline docs/reference/run.md]: @moxiegirl I won't have time for a couple days to update this; if you could carry the doc changes that would be excellent

**Comment by moxiegirl** [comment]
@mavenugo I'll LGTM and carry the change after the merge if you are cool with it.

File: docs/reference/run.md (hunk 5 above)
Comment (mavenugo) [inline docs/reference/run.md]: @dkjer shall I suggest a minor tweak to this Doc and get it merged ?
@moxiegirl can update the doc in a later commit with the table structure that she proposed ?

The proposed documentation is confusing 

> When specifying a range for hostPort only, the containerPort must not be a range.

Confusing and it is conflicting with the other option of providing both hostPort range and containerPort range.

> In this case the container port is published somewhere within the specified hostPort range.

I would reword it as `In this case the container port is mapped to a host port somewhere within the specified hostPort range``

Comment (moxiegirl) [inline docs/reference/run.md]: @mavenugo I'll pick up your change wording when I carry the docs.

**Comment by thaJeztah** [comment]
sgtm, agreed to update after merging so LGTM

@moxiegirl I'll merge this one, then I'll rebase https://github.com/docker/docker/pull/15675 when I'm home, does that work for you?

**Comment by dkjer** [comment]
@moxiegirl Thanks for carrying the doc changes!

**Comment by thaJeztah** [comment]
Thank *you* @dkjer for contributing, and being so patient!
